### PR TITLE
Refactor left nav tree

### PR DESCRIPTION
**Pull request by smadbe**
Get rid of NavMenuItem and make the left nav tree agnostic of the type of the node it is working with.

The left menu should not be impacted by these changed. I've tried to verify everything but I may have miss a few things. 

(unfortunately, as the tree uses templates and template variables are `any`, there is not even type checking for verifying that I am not using a key which does not even exist anymore)

### CODE REVIEW

File: src/app/core/components/left-nav-tree/left-nav-tree.component.html [hunk 1]
```diff
@@ -180,24 +168,23 @@
...
+          <i class="fa fa-user" *ngIf="node.data.type === 'User'"></i>
         </span>
         <span class="node-label">
           <span class="label-container">
             <span class="node-label-title">
               {{ node.label }}
             </span>
-            <span class="node-state" *ngIf="node.data.element.type !== 'User'">
-              <alg-loading class="no-padding" size="medium" *ngIf="node.data.status === 'loading'"></alg-loading>
+            <span class="node-state">
```
[the remark this answers is not in the document]
Comment (smadbe) [reply src/app/core/components/left-nav-tree/left-nav-tree.component.html:179]: Correct. There are no users in the left menu anymore.

File: src/app/core/components/left-nav/left-nav-group-datasource.ts [hunk 2]
```diff
@@ -1,64 +1,72 @@
...
   }
 
-}
+  private mapChild(child: GroupNavigationChild): NavTreeElement {
+    return {
+      id: child.id,
+      title: child.name,
+      hasChildren: child.type !== 'User',
+      navigateTo: (path): void => this.groupRouter.navigateTo(groupRoute({ id: child.id, isUser: false }, path)),
+      groupRelation: { isMember: child.currentUserMembership !== 'none', isManager: mapMananagership(child.currentUserManagership) }
```
[the remark this answers is not in the document]
Comment (smadbe) [reply src/app/core/components/left-nav/left-nav-group-datasource.ts:50]: Ok, I changed it to "managership" keeping the same union type and the backend service. Should be good for now.

File: src/app/core/components/left-nav/left-nav-item-datasource.ts [hunk 3]
```diff
@@ -1,103 +1,106 @@
...
   }
 
-  fetchRootTreeData(): Observable<NavMenuItem[]> {
+  fetchRootTreeData(): Observable<NavTreeElement[]> {
     return this.itemNavService.getRoots(this.category).pipe(
       map(groups => groups.map(g => ({
-        ...mapChild(g.item),
-        groupName: g.name,
+        ...this.mapChild(g.item, defaultAttemptId),
+        associatedGroup: g.name,
```
[the remark this answers is not in the document]
Comment (smadbe) [reply src/app/core/components/left-nav/left-nav-item-datasource.ts:25]: fixed

File: src/app/shared/routing/item-route.ts [hunk 4]
```diff
@@ -47,6 +47,11 @@ export function itemRoute(contentType: ItemTypeCategory, id: ItemId, path: strin
   return { ...rawItemRoute(contentType, id), path };
 }
 
+type MandatoryAttempt = { attemptId: AttemptId } | { parentAttemptId: AttemptId };
+export function fullItemRoute(contentType: ItemTypeCategory, id: ItemId, path: string[], attempt: MandatoryAttempt): FullItemRoute {
```
[the remark this answers is not in the document]
Comment (smadbe) [reply src/app/shared/routing/item-route.ts:51]: Not really "instead", but I can use it in it. Done

File: src/app/core/components/left-nav-tree/left-nav-tree.component.html [hunk 5]
```diff
@@ -227,10 +214,10 @@
       <span class="node-item-content" (click)="selectNode(node)">
         <span class="node-select">
           <alg-score-ring
-            *ngIf="node.data.element.currentScore >= 0; else leafNoScore"
-            [currentScore]="node.data.element.currentScore"
-            [bestScore]="node.data.element.bestScore"
-            [isValidated]="node.data.element.validated"
+            *ngIf="node.data.score !== undefined; else leafNoScore"
```
[the remark this answers is not in the document]
Comment (smadbe) [reply src/app/core/components/left-nav-tree/left-nav-tree.component.html:217]: Well spotted. Without type-checking, it is very hard not to forget anything. :-(
Fixed

File: src/app/core/components/left-nav-tree/left-nav-tree.component.html [hunk 6]
```diff
@@ -19,9 +19,9 @@
   <!-- "chapter" type -->
```
[the remark this answers is not in the document]
Comment (smadbe) [reply src/app/core/components/left-nav-tree/left-nav-tree.component.html:19]: They were two different items. Fixed with #720 , isn't it?

[the remark this answers is not in the document]
Comment (smadbe) [reply src/app/core/components/left-nav-tree/left-nav-tree.component.html:19]: I cannot reproduce the difference (except your two links are not from the same page). Maybe fixed with #720. 
On which DFS do you click? It appears at 3 places. 

I don't think this is what you mention, but there is a known bug (but already before): if you click, in the left menu, on a child which is as well a sibling, the left menu updates to the siblings (not matching the path).

[the remark this answers is not in the document]
Comment (ThomasJuster) [reply src/app/core/components/left-nav-tree/left-nav-tree.component.html:19]: I just checked in production, there's no difference now between prod and branch. But if the bug is known and will be addressed, it's not the scope of this PR then.
Resolving it.

[the remark this answers is not in the document]
Comment (ThomasJuster) [reply src/app/core/components/left-nav-tree/left-nav-tree.component.html:19]: OK my bad, I didn't keep up.
Resolving :+1:

**Review by ThomasJuster — APPROVED** [review]

File: src/app/core/components/left-nav-tree/left-nav-tree.component.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (smadbe) [reply src/app/core/components/left-nav-tree/left-nav-tree.component.html:19]: For the records: the [corresponding issue](https://github.com/France-ioi/AlgoreaFrontend/issues/726)

**Pull request merged**